src/renderer.cpp

```diff
@@ -597,24 +645,140 @@ void Renderer::shutdown()
+    worldMatrix._43 = walkMeshPosOffset;
+
+    struct matrix viewMatrix;
+    memcpy(&viewMatrix.m[0][0], internalState.viewMatrix, sizeof(float) * 16);
+
+    struct matrix worldViewMatrix;
+    multiply_matrix(&worldMatrix, &viewMatrix, &worldViewMatrix);
+
+    memcpy(internalState.worldViewMatrix, &worldViewMatrix.m[0][0], sizeof(float) * 16);
+    internalState.bIsTLVertex = false;
+    internalState.blendMode = RendererBlendMode::BLEND_NONE;
+    internalState.bIsFBTexture = false;
+    internalState.bIsTexture = false;
+
+    setCommonUniforms();
+
+    // Bind shadow map with comparison sampler
+    bgfx::setTexture(3, getUniform(shaderTextureBindings[3], bgfx::UniformType::Sampler), bgfx::getTexture(shadowMapFrameBuffer));
+
+    // Bind shadow map for direct depth sampling
+    uint32_t flags = BGFX_TEXTURE_RT | BGFX_SAMPLER_POINT | BGFX_SAMPLER_U_CLAMP | BGFX_SAMPLER_V_CLAMP | BGFX_SAMPLER_W_CLAMP;
+    bgfx::setTexture(4, getUniform(shaderTextureBindings[4], bgfx::UniformType::Sampler), bgfx::getTexture(shadowMapFrameBuffer), flags);
+
+    // Set state
+    {
+        internalState.state = BGFX_STATE_WRITE_RGB | BGFX_STATE_WRITE_A;
 
-        bgfx::setViewRect(backendViewId, 0, 0, framebufferWidth, framebufferHeight);
+        internalState.state |= BGFX_STATE_CULL_CCW;
 
-        if (internalState.bDoScissorTest) bgfx::setScissor(scissorOffsetX, scissorOffsetY, scissorWidth, scissorHeight);
+        internalState.state |= BGFX_STATE_BLEND_EQUATION(BGFX_STATE_BLEND_EQUATION_ADD);
+        internalState.state |= BGFX_STATE_BLEND_FUNC(BGFX_STATE_BLEND_SRC_ALPHA, BGFX_STATE_BLEND_INV_SRC_ALPHA);
+
+        internalState.state |= BGFX_STATE_DEPTH_TEST_LEQUAL;
+        internalState.state |= BGFX_STATE_WRITE_Z;
     }
+    bgfx::setState(internalState.state);
+
+    uint32_t stencilState = BGFX_STENCIL_TEST_EQUAL;
+    stencilState |= BGFX_STENCIL_FUNC_REF(255);
+    stencilState |= BGFX_STENCIL_FUNC_RMASK(0xff);
+    stencilState |= BGFX_STENCIL_OP_FAIL_S_KEEP;
+    stencilState |= BGFX_STENCIL_OP_FAIL_Z_KEEP;
+    stencilState |= BGFX_STENCIL_OP_PASS_Z_KEEP;
+    bgfx::setStencil(stencilState, stencilState);
+
+    bgfx::submit(VIEW_ID_SCENE, fieldShadowProgramHandle);
+}
+
+void Renderer::draw()
+{
+    if (trace_all || trace_renderer) trace("Renderer::%s with backendProgram %d\n", __func__, backendProgram);   
+
+    // Set view to render in the framebuffer
+    bgfx::setViewFrameBuffer(VIEW_ID_SCENE, backendFrameBuffer);
+
+    // Set current view rect
+    bgfx::setViewRect(VIEW_ID_SCENE, 0, 0, framebufferWidth, framebufferHeight);
+
+    if (internalState.bDoScissorTest) bgfx::setScissor(scissorOffsetX, scissorOffsetY, scissorWidth, scissorHeight);    
 
     // Set current view transform
-    bgfx::setViewTransform(backendViewId, NULL, internalState.backendProjMatrix);
+    bgfx::setViewTransform(VIEW_ID_SCENE, internalState.viewMatrix, internalState.backendProjMatrix);

```

Comment:
The idea of having a variable named `backendViewId` is also to highlight it's actually changing during runtime, while the notation `VIEW_ID_SCENE` seems to be a constant which is definitely not the case, as depending on which game we run (FF7 or FF8) and how many blit commands are required per-frame, the ID might change.

May I ask why this change?

Reply:
The reason for this change is that when using `backendViewId` variable to specify the view id it was very confusing for me to know what view id was used at any time.
For the changes in this PR I needed to draw the shadow map in a view id lower than normal scene rendering because the shadow map is being referred as a texture during the normal scene rendering.
I think that making the view ids constant in an enum makes it clear to which view id is being drawn. By making view ids we can assure that the shadow map pass is always view id 0, normal scene rendering is 1, and posteffect is 2. I think this also makes it clear when debugging with RenderDoc.

That said, I did not know that multiple blit commands could happen in a single frame. I thought that was only used for the swirl effect in ff7 but I do not currently own a copy of ff8 so I cannot currently check if something is broken there (maybe I'll buy a copy this weekend if necessary).

I was wondering, do blit commands need to happen in a lower view id than normal scene rendering?
If not necessary maybe blit commands could start at a constant view id and have a `blitViewId` variable that increases after each blit command. Something like this:
enum VIEW_ID
{
    VIEW_ID_SHADOWMAP,
    VIEW_ID_SCENE,
    VIEW_ID_POSTEFFECT,
    VIEW_ID_BLIT_START,
};

// At start of each frame
blitViewId = VIEW_ID_BLIT_START;

void blitTexture(...)
{
     ...
     bgfx::blit(blitViewId, ...);
     ++blitViewId;
}

This would enable to have constant view ids that make it clear the purpose of each draw call as well as allowing variable view ids for the blit commands where it is necessary. What do you think?

If the above does not look good to you I am ok reverting to using `backendViewId`.

Reply:
Thanks for clarifying this, I get the complexity behind but you can easily workaround that but simply injecting your own code in the `::draw()` call without needing to revamp the overall view ID logic. Just create your own `::drawShadow()` method and call it in the first function body of the `::draw()` method.

Regarding mixing constant view ids with variable ones I think it will be bad because everytime you shift the viewID you need to shift also the constants value. In FF7 blits are used only for the swirl, but in FF8 they're used to optimize some draw calls because in the PSX era it was nearly impossible to render that 3D effects everytime from scratch ( the game would have run half the FPS speed if not less ), so they came up with the concept of blitting the last rendered frame and reusing it, if the next draw call allows it.

Because of this, I'd prefer keeping the current renderer code the way it is, and not touching this logic as it has been battle tested on FF8, and I think it's one of the most delicate pillars of the rendering class itself.

Thanks and sorry :)

Reply:
Thank you for the explanation.

> so they came up with the concept of blitting the last rendered frame and reusing it, if the next draw call allows it.
This seems to imply that the result of blit operations is only used in the next frame. Would this not mean that it does not matter if blit operations are done last and allow the first view ids to be constant?

Anyway, I have reverted the view ID logic to use `backendViewId` variable. It is now exactly as before with just one difference. The `backendViewId` variable now starts at view ID 1 instead of 0. This is because as explained before, I need the shadow map to be drawn in a view ID with smaller value to be able to reference it as a texture during normal rendering. Please let me know what you think.

Reply:
By the way, I said I may buy FF8 to check that nothing is broken but unfortunately it is region locked, meaning that it is not available for purchase or activation in my country :(